models/DataObject/ClassDefinition/Data/Password.php

```diff
@@ -105,27 +105,29 @@ public function getDataForResource(mixed $data, DataObject\Concrete $object = nu
         }
 
         // is already a hashed string? Then do not re-hash
-        $info = password_get_info($data);
-        if ($info['algo'] !== null && $info['algo'] !== 0) {
-            return $data;
-        }
-
-        // password_get_info() will not detect older, less secure, hashing algos.
-        // It might not detect some less common ones as well.
-        $maybeHash = preg_match('/^[a-f0-9]{32,}$/i', $data);
-        $hashLenghts = [
-            32,  // MD2, MD4, MD5, RIPEMD-128, Snefru 128, Tiger/128, HAVAL128
-            40,  // SHA-1, HAS-160, RIPEMD-160, Tiger/160, HAVAL160
-            48,  // Tiger/192, HAVAL192
-            56,  // SHA-224, HAVAL224
-            64,  // SHA-256, BLAKE-256, GOST, GOST CryptoPro, HAVAL256, RIPEMD-256, Snefru 256
-            96,  // SHA-384
-            128, // SHA-512, BLAKE-512, SWIFFT
-        ];
-
-        if ($maybeHash && in_array(strlen($data), $hashLenghts, true)) {
-            // Probably already a hashed string
-            return $data;
+        if($this->getAlgorithm() === self::HASH_FUNCTION_PASSWORD_HASH) {
+            $info = password_get_info($data);
+            if ($info['algo'] !== null && $info['algo'] !== 0) {
+                return $data;
+            }
+        } else {
+            // password_get_info() will not detect older, less secure, hashing algos.
+            // It might not detect some less common ones as well.
+            $maybeHash = preg_match('/^[a-f0-9]{32,}$/i', $data);
+            $hashLenghts = [
```

Comment:
LGTM as it reduces the false positive but casually found another problem when reviewing this area:

we are actually not supporting all the [hash_algos()](https://github.com/pimcore/admin-ui-classic-bundle/blob/d9aff7c5f8378b6bf9f30f1a4b672d48f4198e47/src/Controller/Admin/SettingsController.php#L1404) despite allowing to be selected and used

If i pick one of the following algos that is marked as `not supported` in the screenshot below (eg, `xxh32` which has 8 characters), the character  lenght of some hash is below 32, and when is not matched, it lead to  `$data` to be hashed again
![image](https://github.com/pimcore/pimcore/assets/6014195/c07f1a33-ce46-461c-9d85-086006e02bb5)

(here the quick snippet i just made to check)
```
<?php

        $hashLenghts = [
            32,  // D2, MD4, MD5, RIPEMD-128, Snefru 128, Tiger/128, HAVAL128
            40,  // SHA-1, HAS-160, RIPEMD-160, Tiger/160, HAVAL160
            48,  // Tiger/192, HAVAL192
            56,  // SHA-224, HAVAL224
            64,  // SHA-256, BLAKE-256, GOST, GOST CryptoPro, HAVAL256, RIPEMD-256, Snefru 256
            96,  // SHA-384
            128, // SHA-512, BLAKE-512, SWIFFT
        ];

$algos = hash_algos();

foreach ($algos as $algo){
	echo $algo.' ----- ';
	$hashed = hash($algo, 'test');
	$len = strlen($hashed);
	
	echo $len . ' ---- ';
	if (!in_array($len, $hashLenghts)){
		echo '<b>not supported</b>';	
	}
				  
	echo '<br>';
}
```

Also, the fallback for the Alghos supported by `password_hash` might not be covered in the `ELSE` part (when not matching `self::HASH_FUNCTION_PASSWORD_HASH` but still obtainable by `password_get_info`), 
https://www.php.net/manual/en/function.password-hash.php#:~:text=are%20currently%20supported%3A-,PASSWORD_DEFAULT%20%2D%20Use%20the%20bcrypt%20algorithm%20(default%20as%20of%20PHP%205.5.0).%20Note,is%20only%20available%20if%20PHP%20has%20been%20compiled%20with%20Argon2%20support.,-Supported%20options%20for

Reply:
probably more a task for admin ui bundle to remove the unsupported ones by unsetting them from the array
by looking at this comment `// password_get_info() will not detect older, less secure, hashing algos.`

So only `ripemd320`(80) is likely not working

Reply:
What if i have some users saving password on `hash` but i later change to use `password_hash`? i think the fallback was meant to cover this case as well?

Reply:
@kingjia90 No, not that I can remember - it was never designed to change the hash algo. 
@schmidt-travixmedia Yes, saving the algo along with the hash itself would solve the issue - the problem is that this doesn't cover existing data. 

So in the end, I'd leave the rest now as it is, and remove support for all other algos in v12.

Reply:
> @kingjia90 No, not that I can remember - it was never designed to change the hash algo. @schmidt-travixmedia Yes, saving the algo along with the hash itself would solve the issue - the problem is that this doesn't cover existing data.
> 
> So in the end, I'd leave the rest now as it is, and remove support for all other algos in v12.

A migration script could be provided that detects the used algorithm of a password using PHP's password_get_info() and set the algorithm name in a new property.  Of course, this would have to be done for all Data Objects that contain password fields in a project.

Reply:
@schmidt-travixmedia the problem with the migration would be that `password_get_info` would only get `PASSWORD_BCRYPT` or `PASSWORD_ARGON_*` and still wouldn't retroactively detect the `hash()` algos

What about having 2 separate text fields, one for stored hash as is (for select all, copy/paste/edit stuff on hashed string), one "empty" for inserting a string that will be hashed (which activates once is a dirty field), in both the UI and APIs (getter/setter), and replaces the stored hash.
 
In this way, you always hash it once before persisting, the unhashed string should be never saved anywhere.